File: packages/mcp-server/src/tools/memos.ts
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { pmApi, PmApiError } from '../pm-api.js';

function err(msg: string) { return { content: [{ type: 'text' as const, text: `Error: ${msg}` }] }; }
function ok(data: unknown) { return { content: [{ type: 'text' as const, text: JSON.stringify(data, null, 2) }] }; }
function handleError(e: unknown) { return err(e instanceof PmApiError ? e.message : String(e)); }

export function registerMemosTools(server: McpServer) {
  server.tool('list_memos', 'List memos', {
    project_id: z.string().optional().describe('Project ID'),
    assigned_to: z.string().optional().describe('Filter by assigned team member ID'),
    status: z.string().optional().describe('Filter by status (open/resolved)'),
    q: z.string().optional().describe('Search query'),
    include_archived: z.boolean().optional().describe('Include archived memos (default: false)'),
  }, async ({ project_id, assigned_to, status, q, include_archived }) => {
    try {
      const params = new URLSearchParams();
      if (project_id) params.set('project_id', project_id);
      if (assigned_to) params.set('assigned_to', assigned_to);
      if (status) params.set('status', status);
      if (q) params.set('q', q);
      if (include_archived) params.set('include_archived', 'true');
      const qs = params.toString();
      const data = await pmApi(`/api/v2/memos${qs ? `?${qs}` : ''}`);
      return ok(data);
    } catch (e) { return handleError(e); }
  });

  server.tool('create_memo', 'Create memo', {
    project_id: z.string().optional(),
    title: z.string().optional(),
    content: z.string(),
    memo_type: z.string().optional(),
    assigned_to: z.string().optional().describe('Team member ID to assign'),
    story_id: z.string().optional(),
  }, async (body) => {
    try {
      const data = await pmApi('/api/v2/memos', { method: 'POST', body: JSON.stringify(body) });
      return ok(data);
    } catch (e) { return handleError(e); }
  });

  // [DEPRECATED] send_memo — /api/v2/memos POST로 라우팅 (웹훅 발송 경로 복원, S-B2 버그 수정).
  // 기존 에이전트 호환성 유지. 스프린터블 운영이 send_memo 없이 검증된 후 제거 예정.
  server.tool('send_memo', '[DEPRECATED] Send a memo. Routes to /api/v2/memos which fires webhooks to assigned members. Will be removed after memo-free sprint operation is validated.', {
    project_id: z.string().optional(),
    title: z.string().optional(),
    content: z.string(),
    memo_type: z.string().optional(),
    assigned_to: z.string().optional().describe('Single team member ID (legacy, use assigned_to_ids for multiple)'),
    assigned_to_ids: z.array(z.string()).optional().describe('Team member IDs to assign (supports multiple assignees)'),
    trigger_type: z.string().optional().describe('Workflow stage (kickoff, qa_request, review, merge_request)'),
  }, async ({ assigned_to, assigned_to_ids, project_id, content, title, memo_type, trigger_type }) => {
    try {
      const body: Record<string, unknown> = {
        content,
        title: title ?? '(memo)',
        project_id: project_id ?? '',
        assigned_to,
        assigned_to_ids,
        memo_type,
      };
      if (trigger_type) {
        body.memo_metadata = { trigger_type };
      }
      const data = await pmApi('/api/v2/memos', { method: 'POST', body: JSON.stringify(body) }) as Record<string, unknown>;
      return ok({ ...data, deprecated: true, trigger_type });
    } catch (e) { return handleError(e); }
  });

  server.tool('list_my_memos', 'List memos assigned to or created by a member', {
    assigned_to: z.string().optional().describe('Filter by assigned team member ID'),
    created_by: z.string().optional().describe('Filter by creator team member ID'),
    project_id: z.string().optional(),
    status: z.string().optional(),
  }, async ({ assigned_to, created_by, project_id, status }) => {
    try {
      const params = new URLSearchParams();
      if (project_id) params.set('project_id', project_id);
      if (assigned_to) params.set('assigned_to', assigned_to);
      if (created_by) params.set('created_by', created_by);
      if (status) params.set('status', status);
      const qs = params.toString();
      const data = await pmApi(`/api/v2/memos${qs ? `?${qs}` : ''}`);
      return ok(data);
    } catch (e) { return handleError(e); }
  });

  // AC6: read_memo — conversation 우선, 없으면 memos fallback
  server.tool('read_memo', 'Read memo with replies (reads from conversation if migrated)', {
    memo_id: z.string().describe('Memo ID (also accepts conversation ID)'),
  }, async ({ memo_id }) => {
    try {
      // conversation 존재 여부 확인
      try {
        const msgs = await pmApi(`/api/v2/conversations/${encodeURIComponent(memo_id)}/messages`) as Record<string, unknown>;
        const convData = { id: memo_id, conversation_id: memo_id, replies: msgs };
        return ok(convData);
      } catch {
        // fallback to memos API
      }
      const data = await pmApi(`/api/v2/memos/${encodeURIComponent(memo_id)}`);
      return ok(data);
    } catch (e) { return handleError(e); }
  });

  // [DEPRECATED] reply_memo — 내부적으로 conversation thread reply로 라우팅됨 (S-B2).
  server.tool('reply_memo', '[DEPRECATED] Reply to a memo. Internally routes to conversation thread reply. Will be removed after memo-free sprint operation is validated.', {
    memo_id: z.string(),
    content: z.string(),
    assigned_to: z.string().optional().describe('Single team member ID to explicitly notify via webhook (legacy, use assigned_to_ids for multiple)'),
    assigned_to_ids: z.array(z.string()).optional().describe('Team member IDs to explicitly notify via webhook on this reply'),
  }, async ({ memo_id, content, assigned_to, assigned_to_ids }) => {
    try {
      // AC4: memo_id = conversation_id (S-B1 패턴). top-level root message 조회 → thread reply 생성
      let rootMsgId: string | undefined;
      try {
        // pmApi already unwraps {data: [...]} → returns array directly
        const msgs = await pmApi(`/api/v2/conversations/${encodeURIComponent(memo_id)}/messages?limit=1`) as Array<Record<string, unknown>>;
        rootMsgId = msgs[0]?.id as string | undefined;
      } catch {
        // conversation 없음 → 기존 memos API fallback
      }

      if (rootMsgId) {
        // conversation thread reply
        const resolvedIds = assigned_to_ids ?? (assigned_to ? [assigned_to] : undefined);
        const msgPayload: Record<string, unknown> = { content, thread_id: rootMsgId };
        if (resolvedIds) msgPayload.mentioned_ids = resolvedIds;
        const msg = await pmApi(`/api/v2/conversations/${encodeURIComponent(memo_id)}/messages`, {
          method: 'POST', body: JSON.stringify(msgPayload),
        }) as Record<string, unknown>;
        const msgData = (msg.data ?? msg) as Record<string, unknown>;
        return ok({ memo_id, conversation_id: memo_id, message_id: msgData.id, deprecated: true });
      }

      // fallback: 기존 memos reply API
      const resolvedIds = assigned_to_ids ?? (assigned_to ? [assigned_to] : undefined);
      const payload: Record<string, unknown> = { content };
      if (resolvedIds) payload.assigned_to_ids = resolvedIds;
      const data = await pmApi(`/api/v2/memos/${encodeURIComponent(memo_id)}/replies`, {
        method: 'POST',
        body: JSON.stringify(payload),
      });
      return ok(data);
    } catch (e) { return handleError(e); }
  });

  server.tool('resolve_memo', 'Resolve a memo', {
    memo_id: z.string(),
  }, async ({ memo_id }) => {
    try {
      const data = await pmApi(`/api/v2/memos/${encodeURIComponent(memo_id)}/resolve`, {
        method: 'PATCH',
      });
      return ok(data);
    } catch (e) { return handleError(e); }
  });

  // AC7: send_chat_message — optional thread_id(reply), message_type, review_type, metadata 추가
  server.tool('send_chat_message', 'Send a chat message to a thread (conversation). Triggers real-time SSE delivery to all participants.', {
    thread_id: z.string().describe('Conversation/Thread ID (memo ID = conversation ID)'),
    content: z.string().describe('Message content'),
    reply_thread_id: z.string().optional().describe('Optional: message ID to reply to within the thread (creates sub-thread)'),
    message_type: z.string().optional().describe('Optional: message type hint (e.g. kickoff, qa, review)'),
    review_type: z.string().optional().describe('Optional: review type (approve, request_changes, comment)'),
    metadata: z.record(z.string(), z.unknown()).optional().describe('Optional: extra metadata attached to the message'),
  }, async ({ thread_id, content, reply_thread_id, message_type, review_type, metadata }) => {
    try {
      const payload: Record<string, unknown> = { content };
      if (reply_thread_id) payload.thread_id = reply_thread_id;
      if (message_type || review_type || metadata) {
        payload.metadata = { ...(metadata ?? {}), ...(message_type ? { message_type } : {}), ...(review_type ? { review_type } : {}) };
      }
      const data = await pmApi(`/api/v2/conversations/${encodeURIComponent(thread_id)}/messages`, {
        method: 'POST',
        body: JSON.stringify(payload),
      });
      return ok(data);
    } catch (e) { return handleError(e); }
  });

  // CB-S11: create_conversation — memo 우회 없이 직접 대화 스레드 생성
  server.tool('create_conversation', 'Create a new conversation thread. Returns conversation_id for use with send_chat_message / list_chat_messages.', {
    participant_ids: z.array(z.string()).describe('Team member IDs to include in the conversation'),
    title: z.string().optional().describe('Optional conversation title'),
    project_id: z.string().optional().describe('Project ID (required by backend; use current project if omitted)'),
  }, async ({ participant_ids, title, project_id }) => {
    try {
      const payload: Record<string, unknown> = {
        type: 'group',
        participant_ids,
        project_id: project_id ?? '',
      };
      if (title) payload.title = title;
      const conv = await pmApi('/api/v2/conversations', { method: 'POST', body: JSON.stringify(payload) }) as Record<string, unknown>;
      return ok({ conversation_id: conv.id as string, ...conv });
    } catch (e) { return handleError(e); }
  });

  // S41: conversations API로 라우팅 전환
  server.tool('list_chat_messages', 'List chat messages in a thread (conversation, chronological order)', {
    thread_id: z.string().describe('Thread ID (memo ID = conversation ID)'),
    limit: z.number().optional().describe('Max messages (default: 30)'),
    before: z.string().optional().describe('Cursor: ISO timestamp, fetch messages before this time'),
  }, async ({ thread_id, limit, before }) => {
    try {
      const params = new URLSearchParams();
      if (limit !== undefined) params.set('limit', String(limit));
      if (before !== undefined) params.set('before', before);
      const data = await pmApi(`/api/v2/conversations/${encodeURIComponent(thread_id)}/messages?${params.toString()}`);
      return ok(data);
    } catch (e) { return handleError(e); }
  });
}

File: backend/sprintable_mcp/tools/memos.py
"""메모+채팅 MCP 도구 (10개)."""
from __future__ import annotations

from mcp.types import TextContent

from ..api_client import client
from ..response import err, ok
from ..schemas import SprintableInput


class ListMemosInput(SprintableInput):
    assigned_to: str | None = None
    status: str | None = None
    q: str | None = None
    include_archived: bool | None = None


class CreateMemoInput(SprintableInput):
    content: str
    title: str | None = None
    memo_type: str | None = None
    assigned_to: str | None = None
    assigned_to_ids: list[str] | None = None
    story_id: str | None = None
    trigger_type: str | None = None


class ListMyMemosInput(SprintableInput):
    assigned_to: str | None = None
    created_by: str | None = None
    status: str | None = None


class MemoIdInput(SprintableInput):
    memo_id: str


class ReplyMemoInput(SprintableInput):
    memo_id: str
    content: str
    assigned_to: str | None = None
    assigned_to_ids: list[str] | None = None


class SendChatInput(SprintableInput):
    thread_id: str
    content: str
    reply_thread_id: str | None = None
    message_type: str | None = None
    review_type: str | None = None
    metadata: dict | None = None


class CreateConversationInput(SprintableInput):
    participant_ids: list[str]
    title: str | None = None


class ListChatMessagesInput(SprintableInput):
    thread_id: str
    limit: int | None = None
    before: str | None = None


async def list_memos(args: ListMemosInput) -> list[TextContent]:
    """메모 목록 조회."""
    params: dict = {"project_id": client.project_id}
    if args.assigned_to:
        params["assigned_to"] = args.assigned_to
    if args.status:
        params["status"] = args.status
    if args.q:
        params["q"] = args.q
    if args.include_archived:
        params["include_archived"] = "true"
    try:
        return ok(await client.get("/api/v2/memos", params=params))
    except Exception as exc:
        return err(str(exc))


async def create_memo(args: CreateMemoInput) -> list[TextContent]:
    """메모 생성."""
    body: dict = {"content": args.content, "project_id": client.project_id}
    if args.title:
        body["title"] = args.title
    if args.memo_type:
        body["memo_type"] = args.memo_type
    if args.assigned_to:
        body["assigned_to"] = args.assigned_to
    if args.assigned_to_ids:
        body["assigned_to_ids"] = args.assigned_to_ids
    if args.story_id:
        body["story_id"] = args.story_id
    if args.trigger_type:
        body["memo_metadata"] = {"trigger_type": args.trigger_type}
    try:
        return ok(await client.post("/api/v2/memos", json=body))
    except Exception as exc:
        return err(str(exc))


async def send_memo(args: CreateMemoInput) -> list[TextContent]:
    """[DEPRECATED] 메모 발송. create_memo와 동일 경로."""
    result = await create_memo(args)
    return result


async def list_my_memos(args: ListMyMemosInput) -> list[TextContent]:
    """내 메모 목록 조회 (담당/작성)."""
    params: dict = {"project_id": client.project_id}
    if args.assigned_to:
        params["assigned_to"] = args.assigned_to
    if args.created_by:
        params["created_by"] = args.created_by
    if args.status:
        params["status"] = args.status
    try:
        return ok(await client.get("/api/v2/memos", params=params))
    except Exception as exc:
        return err(str(exc))


async def read_memo(args: MemoIdInput) -> list[TextContent]:
    """메모 읽기 (conversation 우선, 없으면 memos fallback)."""
    try:
        msgs = await client.get(f"/api/v2/conversations/{args.memo_id}/messages")
        return ok({"id": args.memo_id, "conversation_id": args.memo_id, "replies": msgs})
    except Exception:
        pass
    try:
        return ok(await client.get(f"/api/v2/memos/{args.memo_id}"))
    except Exception as exc:
        return err(str(exc))


async def reply_memo(args: ReplyMemoInput) -> list[TextContent]:
    """[DEPRECATED] 메모 답신 (conversation thread reply로 라우팅)."""
    try:
        msgs = await client.get(f"/api/v2/conversations/{args.memo_id}/messages", params={"limit": "1"})
        root_msgs = msgs if isinstance(msgs, list) else (msgs.get("data") or [])
        root_msg_id = root_msgs[0].get("id") if root_msgs else None

        if root_msg_id:
            resolved_ids = args.assigned_to_ids or ([args.assigned_to] if args.assigned_to else None)
            payload: dict = {"content": args.content, "thread_id": root_msg_id}
            if resolved_ids:
                payload["mentioned_ids"] = resolved_ids
            msg = await client.post(f"/api/v2/conversations/{args.memo_id}/messages", json=payload)
            msg_data = msg if not isinstance(msg, dict) or "data" not in msg else msg["data"]
            return ok({"memo_id": args.memo_id, "conversation_id": args.memo_id, "message_id": msg_data.get("id") if isinstance(msg_data, dict) else None, "deprecated": True})
    except Exception:
        pass

    try:
        resolved_ids = args.assigned_to_ids or ([args.assigned_to] if args.assigned_to else None)
        payload = {"content": args.content}
        if resolved_ids:
            payload["assigned_to_ids"] = resolved_ids
        return ok(await client.post(f"/api/v2/memos/{args.memo_id}/replies", json=payload))
    except Exception as exc:
        return err(str(exc))


async def resolve_memo(args: MemoIdInput) -> list[TextContent]:
    """메모 해결 처리."""
    try:
        return ok(await client.patch(f"/api/v2/memos/{args.memo_id}/resolve"))
    except Exception as exc:
        return err(str(exc))


async def send_chat_message(args: SendChatInput) -> list[TextContent]:
    """conversation thread에 채팅 메시지 발송."""
    payload: dict = {"content": args.content}
    if args.reply_thread_id:
        payload["thread_id"] = args.reply_thread_id
    meta: dict = {}
    if args.metadata:
        meta.update(args.metadata)
    if args.message_type:
        meta["message_type"] = args.message_type
    if args.review_type:
        meta["review_type"] = args.review_type
    if meta:
        payload["metadata"] = meta
    try:
        return ok(await client.post(f"/api/v2/conversations/{args.thread_id}/messages", json=payload))
    except Exception as exc:
        return err(str(exc))


async def create_conversation(args: CreateConversationInput) -> list[TextContent]:
    """새 conversation thread 생성."""
    body: dict = {
        "type": "group",
        "participant_ids": args.participant_ids,
        "project_id": client.project_id,
    }
    if args.title:
        body["title"] = args.title
    try:
        conv = await client.post("/api/v2/conversations", json=body)
        conv_id = conv.get("id") if isinstance(conv, dict) else None
        return ok({"conversation_id": conv_id, **(conv if isinstance(conv, dict) else {})})
    except Exception as exc:
        return err(str(exc))


async def list_chat_messages(args: ListChatMessagesInput) -> list[TextContent]:
    """conversation thread 메시지 목록 조회."""
    params: dict = {}
    if args.limit is not None:
        params["limit"] = str(args.limit)
    if args.before:
        params["before"] = args.before
    try:
        return ok(await client.get(f"/api/v2/conversations/{args.thread_id}/messages", params=params))
    except Exception as exc:
        return err(str(exc))
